### PR TITLE
Wip/hughsie/threading

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: src/meson.build [hunk 1]
```diff
@@ -303,6 +306,7 @@ if get_option('tests')
       '-DTESTDATADIR="' + testdatadir_src + ':' + testdatadir_dst + '"',
       '-DPLUGINBUILDDIR="' + pluginbuilddir + '"',
       '-DFU_OFFLINE_DESTDIR="/tmp/fwupd-self-test"',
+      '-DFU_MUTEX_DEBUG',
```
Comment (superm1) [inline src/meson.build]: This reminds me that some other aspects of the daemon are too verbose when called `--verbose`.  For example the `"using ------------- for $str"` and dispatching into plugins for usb and udev devices.  What do you think about putting those behind #defines as well?  Or having a concept of a daemon verbosity level?

File: src/fu-device.c [hunk 2]
```diff
@@ -1904,10 +1925,14 @@ fu_device_incorporate (FuDevice *self, FuDevice *donor)
 	FuDevicePrivate *priv = GET_PRIVATE (self);
 	FuDevicePrivate *priv_donor = GET_PRIVATE (donor);
 	GPtrArray *parent_guids = fu_device_get_parent_guids (donor);
+	g_autoptr(FuMutexLocker) locker3 = fu_mutex_read_locker_new (priv_donor->metadata_mutex);
+	g_autoptr(FuMutexLocker) locker4 = fu_mutex_read_locker_new (priv_donor->parent_guids_mutex);
```
Comment (superm1) [inline src/fu-device.c]: Don't really like just adding on numeric lockers here.  `donor_metadata_locker` and `donor_parent_locker` maybe?

**Comment by superm1** [comment]
Oh I didn't think you were going to get the threading in for 1.1.2.
I think you'll need to put mutex protection around remotes as well.

File: src/meson.build (hunk 1 above)
Comment (hughsie) [reply src/meson.build]: Well, how verbose is too verbose? It's too verbose when there's so much information you can't find the issue, but it's not verbose enough you can't pinpoint a failure. I don't think warning "levels" help much, but I do think we could separate this into various "classes" of debugging that can be enabled on end-user systems using either a cli option or env vars.

File: src/fu-device.c (hunk 2 above)
Comment (hughsie) [reply src/fu-device.c]: Right, sorry, this was forgotten about.

**Comment by hughsie** [comment]
> Oh I didn't think you were going to get the threading in for 1.1.2.

No, it was just me playing with half a brain recovering from food poisoning. I was tempted to roll 1.1.2 with what's in master now, and then do a follow up 1.1.3 a week later when the inevitable issues roll in. Threading can wait a long time.

File: src/meson.build (hunk 1 above)
Comment (superm1) [reply src/meson.build]: Yeah classes of verbosity levels is really what I had in mind.  Perhaps making `--verbose` have an optional argument to remain compatible with previous behavior (no argument default to `all`)

**Comment by superm1** [comment]
> I was tempted to roll 1.1.2 with what's in master now,

Sounds good to me.

File: src/meson.build (hunk 1 above)
Comment (hughsie) [reply src/meson.build]: Some ideas in https://github.com/hughsie/fwupd/pull/738

File: src/fu-mutex.c [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+ * fu_mutex_lock:
+ * @self: a #FuMutex
+ *
+ * Obtain a read or write lock.
+ *
+ * If @kind is FU_MUTEX_ACCESS_READ and another thread currently holds the
+ * write lock or blocks waiting for it, the current thread will block.
+ * Read locks can be taken recursively.
+
+ * If @kind is FU_MUTEX_ACCESS_READ and any thread already holds a read or
```
Comment (gicmo) [inline src/fu-mutex.c]: I guess you mean `FU_MUTEX_ACCESS_WRITE` here?

**Comment by hughsie** [comment]
@superm1 any worries including this threading prep in 1.1.3? There's no actual threads in play yet, but I'd like to get some testing of the protections before working on that bit.

**Comment by superm1** [comment]
No, no concerns

**Pull request merged**